core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -321,14 +321,14 @@ class Partition(val topicPartition: TopicPartition,
   // from its partitionStates if this method returns true
   def maybeReplaceCurrentWithFutureReplica(): Boolean = {
     val replica = localReplicaOrException
-    val futureReplicaLEO = futureLocalReplica.map(_.logEndOffset)
-    if (futureReplicaLEO.contains(replica.logEndOffset)) {
+    val futureReplicaLEO = futureLocalReplica.map(_.logEndOffset.messageOffset)
+    if (futureReplicaLEO.contains(replica.logEndOffset.messageOffset)) {
       // The write lock is needed to make sure that while ReplicaAlterDirThread checks the LEO of the
       // current replica, no other thread can update LEO of the current replica via log truncation or log append operation.
       inWriteLock(leaderIsrUpdateLock) {
         futureLocalReplica match {
           case Some(futureReplica) =>
-            if (replica.logEndOffset == futureReplica.logEndOffset) {
+            if (replica.logEndOffset.messageOffset == futureReplica.logEndOffset.messageOffset) {
```

Comment:
One thing I was discussing with @apovzner is whether we should consider renaming `logEndOffset`. The bug probably would never have occurred if we had a better name. Maybe we should have two methods:

```scala
def logEndOffset: Long
def logEndOffsetMetadata: LogOffsetMetadata
```

Reply:
I like the suggestion of having `logEndOffsetMetadata` method. Although, need to be careful not to miss any calls that would need to be renamed to  
logEndOffsetMetadata (and get to the same type of bug we found here). I guess one way to make sure this does not happen is to first change the name of logEndOffset to 
logEndOffsetMetadata, make sure to compile, and only after that add 
logEndOffset that returns Long.